### DISCUSSION TITLE
[@types/node] Node Test Runner, Export `Mock` type?

### DISCUSSION THREAD

**Discussion by jameslawson** [DefinitelyTyped/DefinitelyTyped#66594] [Issues with a @types package] [answered]
In [test.d.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v18/test.d.ts) I was hoping that the `Mock` type would be exported in order to make this possible? : 

```typescript
// What I would like to write (no type error):
import type { Mock } from 'node:test';

type ConsoleLogMock = Mock<typeof console.log>;
const consoleLogMocked = console.log as ConsoleLogMock;
  
test('mocking console.log', async () => {
  mock.method(console, 'log');
  console.log('hello');
  assert.strictEqual(consoleLogMocked.mock.callCount, 1);
});
```
The `Mock` type is there inside test.d.ts but afaik I can't access it, and without access to `Mock` I don't think it's possible to use mocks in such a type-safe way as Typescript does not know how to type the `.mock` property after console.log has been mocked (but if I missed an alternate way please lmk).

```typescript
// Currently (type error):
test('mocking console.log', () => {
  mock.method(console, 'log');
  console.log('hello');
  
  // Property 'mock' does not exist on type '(message?: any, ...optionalParams: any[]) => void'.ts(2339)
  assert.strictEqual(console.log.mock.callCount(), 1);
  //                             ~~~~
});
```

This idea of exporting a Mock type to allow writing mocks in a more type-safe way is done by many test runners. For example, [Jest](https://jestjs.io/) exports [`jest.Mock<T>`](https://jestjs.io/docs/mock-function-api#jestmockt).

To make this change happen, I think it's as simple as a one-line change in [test.d.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d3a3e808227f8d4d42007440bdd8a6a2c0758b4a/types/node/v18/test.d.ts)?:

```diff
- export { test as default, run, test, describe, it, before, after, beforeEach, afterEach, mock };
+ export { test as default, run, test, describe, it, before, after, beforeEach, afterEach, mock, Mock };
```

[a comment by an automated account is not shown]

**Comment by Semigradsky** [comment]
Looks good for me.

JFYI you can write more type-safely without proposed change:
```ts
test('mocking console.log', async () => {
  const consoleLogMocked = mock.method(console, 'log');
  console.log('hello');
  assert.strictEqual(consoleLogMocked.mock.callCount, 1);
});
```

**Reply by jameslawson** [reply]
Oh cool, thanks for pointing that out I didn't see that 😀

**Comment by Semigradsky** [comment, marked as answer]
PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66768

**Reply by jameslawson** [reply]
Amazing! Thank you so much!
